### ISSUE TITLE
"use_self" false positive when a struct is defined within parent type's method

### ISSUE THREAD

**Issue by aldanor** [rust-lang/rust-clippy#3637]
Clippy would wrongfully recommend to use `Self` in this case (wrong because `Self` now refers to the inner struct):

```rust
impl Foo {
    fn method() {
        struct Bar {
            foo: Foo, //^ warning: unnecessary structure name repetition
        }
    }
}
```
